Backpack/NavigationBar/Classes/BPKNavigationBar.m

```diff
@@ -0,0 +1,291 @@
+/*
+ * Backpack - Skyscanner's Design System
+ *
+ * Copyright 2018-2019 Skyscanner Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#import "BPKNavigationBar.h"
+
+#import <Backpack/Common.h>
+#import <Backpack/Color.h>
+#import <Backpack/Label.h>
+
+#import "BPKNavigationBarLargeTitleView.h"
+#import "BPKNavigationBarTitleView.h"
+
+
+const CGFloat BPKNavigationBarExpandedFullHeight = 96;
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface BPKNavigationBar()
+// Title views
+@property(nonatomic, strong, readonly) BPKNavigationBarLargeTitleView *largeTitleView;
+@property(nonatomic, strong, readonly) BPKNavigationBarTitleView *titleView;
+
+// Background
+@property(nonatomic, strong, readonly) UIVisualEffectView *backgroundView;
+@property(nonatomic, strong, readonly) UIView *borderView;
+
+// Constraints
+@property(nonatomic, strong) NSLayoutConstraint *heightConstraint;
+@property(nonatomic, strong) NSLayoutConstraint *backgroundViewTopConstraint;
+
+/**
+ * Keeps track of the state of the navigation bar. Will be `YES`
+ * when the large title is hidden and the title is displayed in
+ * `titleView`. Otherwise it's `NO`.
+ * The default values is `NO`.
+ */
+@property(nonatomic, assign, getter=isCollapsed) BOOL collapsed;
+@end
+
+
+@implementation BPKNavigationBar
+@synthesize largeTitleView = _largeTitleView,
+            titleView = _titleView,
+            backgroundView = _backgroundView,
+            borderView = _borderView;
+
+- (instancetype)initWithCoder:(NSCoder *)aDecoder {
+    self = [super initWithCoder:aDecoder];
+
+    if (self) {
+        [self setUp];
+    }
+
+    return self;
+}
+
+- (instancetype)initWithFrame:(CGRect)frame {
+    self = [super initWithFrame:frame];
+
+    if (self) {
+        [self setUp];
+    }
+
+    return self;
+}
+
+- (void)setTitle:(NSString *)title {
+    BPKAssertMainThread();
+    if (title != _title) {
+        _title = [title copy];
+
+        self.largeTitleView.titleLabel.text = _title;
+        self.titleView.titleLabel.text = _title;
+    }
+}
+
+- (void)setLargeTitleTextColor:(UIColor *)largeTitleTextColor {
```

Comment:
```suggestion
- (void)setLargeTitleTextColor:(UIColor *_Nullable)largeTitleTextColor {
```

Reply:
I'm surprised that Xcode didn't tell me that

Reply:
I expect if you clean first it would. Anyway - free to merge once you fix that 👍